### ISSUE TITLE
Allow Use of IConfigureOptions Consistently Everywhere

### ISSUE THREAD

**Issue by RehanSaeed** [dotnet/aspnetcore#39251]
### Is your feature request related to a problem? Please describe.
[`IConfigureOptions<T>`](https://andrewlock.net/simplifying-dependency-injection-for-iconfigureoptions-with-the-configureoptions-helper/) allows you to move your configuration code to a separate class. However, it cannot be used with certain API's which require us to pass an `Action<T>`:

```cs
IServiceCollection AddStackExchangeRedisCache(this IServiceCollection services, Action<RedisCacheOptions> setupAction)
IServiceCollection AddHsts(this IServiceCollection services, Action<HstsOptions> configureOptions)
```

This is not an exhaustive list, there may be others.

### Describe the solution you'd like
All API's in ASP.NET Core should be consistent and allow us to configure them using `IConfigureOptions<T>`. The following overloads should be added:

```cs
IServiceCollection AddStackExchangeRedisCache(this IServiceCollection services)
IServiceCollection AddHsts(this IServiceCollection services)
```

This is not an exhaustive list, there may be others.

**Comment by davidfowl** [comment]
They compose, so it can be used. Are you saying you want overloads that don’t require the action overload?

**Comment by RehanSaeed** [comment]
Yes, at the moment it's pretty strange to be forced to pass an empty delegate:

```cs
services.AddStackExchangeRedisCache(x => {});
services.AddHsts(x => {});
```

**Comment by BrennanConroy** [comment]
Today we have two common patterns for configuring options when adding built-in ASP.NET Core features.
1. Single method that takes an `Action<OptionType>?` and defaults to null
https://github.com/dotnet/aspnetcore/blob/c85baf8db0c72ae8e68643029d514b2e737c9fae/src/Components/Server/src/DependencyInjection/ComponentServiceCollectionExtensions.cs#L31
1. Two methods, one without `Action<OptionType>` and the second with `Action<OptionType>`
https://github.com/dotnet/aspnetcore/blob/05b2afb73b55b11d8cb3691dedfebba1f0285332/src/Security/Authorization/Core/src/AuthorizationServiceCollectionExtensions.cs#L21
https://github.com/dotnet/aspnetcore/blob/05b2afb73b55b11d8cb3691dedfebba1f0285332/src/Security/Authorization/Core/src/AuthorizationServiceCollectionExtensions.cs#L47

We should decide on the preferred approach and update any areas that are currently forcing an `Action<OptionType>` when they don't need to be.
Example, `AddCookiePolicy` is fine forcing the `Action<CookiePolicyOptions>` because that's the whole point of the API https://github.com/dotnet/aspnetcore/blob/c85baf8db0c72ae8e68643029d514b2e737c9fae/src/Security/CookiePolicy/src/CookiePolicyServiceCollectionExtensions.cs#L19

Bringing this up with API review so we can agree on the pattern we want to use before updating places like `AddStackExchangeRedisCache`.

**Comment by pranavkm** [comment]
* API review:
* Don't change existing overloads.
* For existing APIs that are missing the overload that does not accept an options callback, we should add one. For e.g we should do this:

```diff
+ IServiceCollection AddStackExchangeRedisCache(this IServiceCollection services)
IServiceCollection AddStackExchangeRedisCache(this IServiceCollection services, Action<RedisCacheOptions> setupAction)

```

* For new ones, we think standardize on using the pattern of having two overloads. This should allow for new options to be introduced in a future release without deviating from the pattern and also conforms to existing patterns in the framework.

```C#
public static IServerSideBlazorBuilder AddMyFeature(this IServiceCollection services);
public static IServerSideBlazorBuilder AddMyFeature(this IServiceCollection services, Action<MyFeatureOption> configureOptions) ;
```

`configureOptions` in the above example can be allowed to be soft-allow null values (i.e. not perform null checks) for the sake of brevity of implementation.

* In the event we only need the options type to configured without adding any other services, we should standardize on `Configure???` as the name for the method e.g. `ConfigureMyFeatureOptions(Action<MyFeatureOptions> configureOptions)` where `configureOptions` is non-null.

**Comment by halter73** [comment]
We decided on the PR (#40689) that the add `AddHsts()` overload without a callback effectively does nothing. [1] Now the new `AddStackExchangeRedisCache` overload has even been called into question. Not because it does nothing, but because redis cannot work without configuring custom options.

Options can be configured separately without the callback, but this is uncommon. We're worried that introducing this overload will create confusion only to save seven characters of code in the uncommon case.

```diff
+ IServiceCollection AddStackExchangeRedisCache(this IServiceCollection services)
```

[1]: This brings up questions about whether `UseHsts` should have taken the options directly similar to `UseStaticFiles` and let people call `.Configure<HstsOptions>(` on the service collection themselves if they are so inclined, but I digress.

**Comment by halter73** [comment]
API Review Notes:

> Options can be configured separately without the callback, but this is uncommon. We're worried that introducing this overload will create confusion only to save seven characters of code in the uncommon case.

We are worried that adding the non-options-configuring `AddStackExchangeRedisCache(this IServiceCollection services)` will create too much confusion by making people think configuring options isn't necessary. If options are configured seperately, calling `AddStackExchangeRedisCache(this IServiceCollection services, Action<RedisCacheOptions> configure)` with `_ => { }` is easy enough.
